Fixes issue #4 and #5

GoldenLin9 opened the pull request:
**What?**
To address issue #4, I fixed the regex not hiding certain characters properly on the player side. For example, "é" in "Qué" before would be displayed to the user when it should be hidden

To address issue #5, I made it so that returning back to a saved game and switching between automatic and manual saves the words that the user chose to have hidden.

**How?**
To fix issue #4, I noticed that there's a function on the player side that reveals the first and/or the last character of a string depending on if that last/first character is a non-latin character. So, "Done!" would appear as "____!", but non-latin letters should also be hidden. To do this, I changed the `notAlphaRegex` regex to match any unix characters not under the category of letters.

To fix issue #5, I noticed that the current game only maintains an array, `hiddenWords`, to store and display the chosen hidden words across different modes ("manual" and "automatic"), but when switching from manual to automatic, `hiddenWords` changes to store the hidden words for automatic and _forgets_ the content of the array from manual. So, I created a temporary array, `previousHiddenWords`, to store the contents of `hiddenWords` when switching to automatic and populate `hiddenWords` with `previousHiddenWords` when the user switches back to manual. I also have a set that holds the indices of all `previousHiddenWords`. Whenever the user saves draft, the indices of the words get saved. When the user opens up the same game, i store the indices to my set and grab all the associated words at those indices. This method can result in an error refrencing the intended chosen hidden words if the user edits the pargraph, so I created a warning label in the edit paragraph page.

**Why?**
- Fixes #4
- Fixes #5

GoldenLin9 committed:
saves hidden words chosen when switching between manual and automatic

GoldenLin9 committed:
properly hides letters beyond latin characters in player side

GoldenLin9 committed:
saves manually chosen words when saving draft and takes user to edit paragraph mode when initializing widget

GoldenLin9 committed:
saves indices no matter what mode the user is currently in

clpetersonucf reviewed (changes requested):
A few things I need you to take another look at:

1. The paragraph warning text on menu 1 should probably only be displayed when a user goes _back_ to edit the paragraph. It's a little redundant to display initially, IMO.
2. Loading an existing qset (via `initExistingWidget`) brings us to menu 2, but the selection appears to default to "Automatic" instead of "Let me choose", even if the latter was selected and saved to the qset.
3. Loading an existing qset does not appear to restore the correct set of hidden words when "Automatic" was selected (the value in "number of words between hidden words" is correct, but the actual words in the list reverts to the default selections, every 3). This isn't just a visual issue - if I then save and preview the widget, the hidden words in the player will be the default frequency (every 3) instead of the frequency I had previously selected.
4. If "Let Me Choose" was selected and a set of hidden words selected, then I go back and edit the paragraph, the full set of previously selected words are still displayed in the list. If I then try to save, I get a JS error about a missing index position, which I assume is because a subset of those words are no longer present. I think the best resolution to this is to simply clear the list of previously hidden words if a change to the paragraph is made, or perhaps only display the hidden words that are still present at their index positions. These appear to be tracked in some capacity since those untouched words will still be highlighted in the paragraph text in menu 2.
5. The "Automatic" vs "Let Me Choose" selection doesn't appear to persist when I transition between menus 1 and 2. Similar to the second problem I mentioned above, I'd like my previous selection to persist not just between saving and reloading the qset, but between transitions between the two menus.

Lastly - this doesn't have to be fixed for this PR, but in the future, you **must make your changes on a branch, not directly on master**. Even if it's a fork - it is very difficult to resolve changes between `ucfopen/master` and `yourfork/master`. Please make sure you first branch off master (or another source branch, like a dev branch) before working on future issues.

GoldenLin9 committed:
Fixes comment 1: warning text displays only when going back to paragraph and not on initial load

GoldenLin9 committed:
Fixes comment 2 & 5: mode state saves between switching menus and initializing an existing widget

GoldenLin9 committed:
Fixes comment 4: updating paragraph updates hidden words at hidden word indices and out of range hidden indices from editing paragraph gets removed

GoldenLin9 committed:
Fixes comment 3: properly saves number of words between a hidden word

GoldenLin9 committed:
adds null check for removePunc function

GoldenLin9 commented:
Fixed! 

Also, for number 4, I decided to update the hidden words at the hidden word indices stored and if an index from my hidden indices is now out of range due to the paragraph being truncated, I removed that index from my set.

clpetersonucf commented on src/creator-logic.coffee:
```diff
@@ -117,33 +145,61 @@ Namespace('Wordguess').CreatorLogic = do ->
 			'paragraph'             : replaceTags(paragraph.join ' ')
 			'wordsToSkip'           : wordsToSkip
 			'manualSkippingIndices' : manualSkippingIndices
+			'mode'				    : currentMode
```
Be **very, very, very** mindful when you change the qset structure here. Consider:

If this is a widget that's already on production (which it is), there are going to be lots of existing qsets that **will not have this structure**, at least until those instances are edited and re-saved (which is never a guarantee). Therefore, if you make a change like this, you must always make sure:

1. The player expects this change in how it references qset data (while _also_ handling old qsets that do not have this change)
2. The creator knows what to do if it loads a qset without this change (is there a default value? Should it look for something in another location in the qset?)

If possible, try and avoid unnecessary qset changes. Work with the existing schema unless you absolutely must change it.

GoldenLin9 replied:
unfortunately, I don't think I can get around not adding another field to the qset structure. The current implementation has a variable, wordsToSkip, that is set to -1 if the manual selection was chosen otherwise a number if automatic selection was selected. Therefore, there is no information in the qset that saves the value for wordsToSkip on the creator side if manual selection was chosen. Therefore, I created another field that will only be used on the creator side storing the value for previous automatic values and if it's not present due to a qset being used from a previous version, I provided a default value

clpetersonucf commented on src/player-logic.coffee:
```diff
@@ -1,5 +1,7 @@
-	regexNotAlpha = /[^A-z]/
+
+	# Regex to match non-letter characters (non-Unicode letters)
+	regexNotAlpha = /[^\p{L}]/u
```
This is great - but you're only changing `regexNotAlpha` in the player! Remember the creator also uses `regexNotAlpha` when it calls `removePunc()`. The old `regexNotAlpha` expression is still being used when the hidden word strings are inserted into the qset in the `answers->text` property.

clpetersonucf reviewed (changes requested):
Two critical changes that must be addressed before this can be QA'd any further.

GoldenLin9 committed:
fixes regexNotAlpha on creator side resulting in wrong intended answers saving

GoldenLin9 committed:
fixes qset structure to work for old and future versions with updated features

GoldenLin9 committed:
adds error text when user doesn't select text to hide on manual mode

GoldenLin9 committed:
invalidates qset if no words chosen to hide

GoldenLin9 committed:
fixes -1 automatic value

clpetersonucf commented on src/creator.coffee:
```diff
@@ -16,21 +16,56 @@ Namespace('Wordguess').Creator = do ->
 	initExistingWidget = (title, widget, qset, version, baseUrl) ->
 		wordsToSkip = qset.wordsToSkip
-		if wordsToSkip is -1 then wordsToSkip = 3
+		previousWordsToSkip = qset.previousWordsToSkip
```
You're doing an undefined check later (on line 32) but I would rather include a fallback here, given `previousWordsToSkip` is another new addition to the qset and may not always exist.

clpetersonucf reviewed (changes requested):
I re-tested again this morning and was unable to duplicate the issue I reported yesterday associated with an index offset. If I had to guess, it might've been browser cache shenanigans.

I requested one final change, and I would just make sure you do a general cleanup pass to remove commented-out code and console logs, and then this looks good to go.

GoldenLin9 committed:
cleanes up code, removes code comments, and better null check

clpetersonucf reviewed (approved):
Tested with old qsets and it appears backwards compatible. Looks good to me, nice job.
